### PR TITLE
Laser auto-on with tablet and Web overlays

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5933/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5933/HighFidelity-Beta-PR12087-6c16fcfe651af9539f80a73a8b16bc12757eab5c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5933/HighFidelity-Beta-PR12087-6c16fcfe651af9539f80a73a8b16bc12757eab5c.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5934/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5934/HighFidelity-Beta-PR12087-6095a90e6972551f47eae10612c5aaa8249db7cc.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5934/HighFidelity-Beta-PR12087-6095a90e6972551f47eae10612c5aaa8249db7cc.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5955/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5955/HighFidelity-Beta-PR12087-27bed8d1f17ceddd13dbcd87118a757926df4e4f.exe

**Comment by ctrlaltdavid** [comment]
**Test Plan**

1. Settings > General > UI > Prefer Avatar Finger Over Stylus.
* Option is no longer present.
* Cannot use index finger to operate tablet any more.

2. Settings > General > UI > Prefer Stylus Over Laser: false
* Lasers (blue) automatically appear when triggers are not squeezed and you orient your hands such that lasers would intersect the tablet.
* Lasers automatically appear / disappear according to their intersection with the tablet model, not just the screen.
* If both hands are pointing at the tablet, only the laser for the domininant hand is drawn. (Settings > .... .)
* If both hands are pointing at the tablet, if you squeeze the trigger for the non-drawn hand the laser changes to that hand.
* false value is remembered over program runs.

3. Settings > General > UI > Prefer Stylus Over Laser: true
* When this is enabled, stylii appear on the hands when near the tablet (same as in master); lasers do not automatically appear.
* Can still use lasers to operate the tablet when you squeeze the controller triggers.
* true value is remembered over program runs.

4. Changing the hand that has the laser on the tablet.
* You can use either hand to work the laser on the tablet.
* If both hands are pointing at the tablet, the dominant hand has the laser. (Settings > Avatar > Dominant Hand.)
* If both hands are pointing at the tablet, you can change the hand that has the laser by (partly or fully) squeezing the trigger on the new hand.

5. Test using the Shapes app.
Either from the Marketplace or from the unpublishedScripts directory from source control.
* Start / stop the app.
* Use the app UI.
* Use the app to manipulate entities.

**Comment by ctrlaltdavid** [comment]
**Code Review**

The FIXMEs added in PreferencesDialog.cpp and Application.h are the subject of https://highfidelity.fogbugz.com/f/cases/11037/Remove-Prefer-Avatar-Finger-Over-Stylus-general-setting-code

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6000/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6000/HighFidelity-Beta-PR12087-69dff9ac09e97d71c78870c191d422a921986fce.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6000/HighFidelity-Beta-PR12087-69dff9ac09e97d71c78870c191d422a921986fce.dmg

File: scripts/system/controllers/controllerModules/inVREditMode.js [hunk 1]
```diff
@@ -19,15 +19,24 @@ Script.include("/~/system/libraries/controllerDispatcherUtils.js");
...
+            100,
+            makeLaserParams(NO_HAND_LASER, false)
         );
 
+        this.pointingAtTablet = function (objectID) {
+            if (objectID === HMD.tabletScreenID || objectID === HMD.homeButtonID) {
+                return true;
+            }
+            return false;
+        };
```
Comment (AndrewMeadows) [inline scripts/system/controllers/controllerModules/inVREditMode.js]: Why not do:
```
this.pointingAtTablet = function (objectID) {
    return objectID === HMD.tabletScreenID || objectID === HMD.homeButtonID;
}
```

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6019/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6019/HighFidelity-Beta-PR12087-5431b7137b2fef5db92882762020f2adf9eed452.dmg

**Comment by ctrlaltdavid** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6022/app_12087_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6022/HighFidelity-Beta-PR12087-5431b7137b2fef5db92882762020f2adf9eed452.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6022/HighFidelity-Beta-PR12087-5431b7137b2fef5db92882762020f2adf9eed452.dmg

**Comment by sbennett77** [comment]
Currently Testing.

**Comment by sbennett77** [comment]
Passed.

**Pull request merged**